arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,600 @@
+      "grid_resolution_width", d->settings.grid_resolution_width );
+  d->settings.grid_resolution_height =
+    config->get_value< unsigned >(
+      "grid_resolution_height", d->settings.grid_resolution_height );
+
+  d->color_resolution = d->color_resolution_per_chan *
+                        d->color_resolution_per_chan *
+                        d->color_resolution_per_chan;
+
+  d->color_histogram.resize( d->color_resolution, 0 );
+  d->intensity_histogram.resize( d->intensity_resolution, 0 );
+}
+
+// ----------------------------------------------------------------------------
+bool
+color_commonality_filter
+::check_configuration( vital::config_block_sptr config ) const
+{
+  auto const color_resolution_per_chan =
+    config->get_value< unsigned >( "color_resolution_per_channel" );
+  auto const intensity_resolution =
+    config->get_value< unsigned >( "intensity_resolution" );
+
+  if( !is_power_of_two( color_resolution_per_chan ) )
+  {
+    LOG_ERROR( logger(), "color_resolution_per_chan must be a power of 2, "
+                         " but instead is: " << color_resolution_per_chan );
+    return false;
+  }
+  if( !is_power_of_two( intensity_resolution ) )
+  {
+    LOG_ERROR( logger(), " intensity_resolution must be a power of 2, but "
+                         "instead is: " << intensity_resolution );
+    return false;
+  }
+  return true;
+}
+
+// ----------------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+color_commonality_filter
+::filter( kwiver::vital::image_container_sptr image_data )
+{
+  // Perform Basic Validation
+  if( !image_data )
+  {
+    return kwiver::vital::image_container_sptr();
+  }
+
+  if( image_data->depth() != 1 && image_data->depth() != 3 )
+  {
+    LOG_ERROR( logger(), "Unsupported number of input planes! Expected 1 or "
+                         "3 but instead was " << image_data->depth() );
+    return kwiver::vital::image_container_sptr();
+  }
+
+  // Get input image
+  vil_image_view_base_sptr view =
+    vxl::image_container::vital_to_vxl( image_data->get_image() );
+
+  // Perform different actions based on input type
+#define HANDLE_CASE( T )                                             \
+  case T:                                                            \
+  {                                                                  \
+    using pix_t = vil_pixel_format_type_of< T >::component_type;     \
+    vil_image_view< pix_t > input = view;                            \
+    return d->compute_commonality( input );                          \
+  }                                                                  \
+
+  switch( view->pixel_format() )
+  {
+    HANDLE_CASE( VIL_PIXEL_FORMAT_BOOL );
+    HANDLE_CASE( VIL_PIXEL_FORMAT_BYTE );
+    HANDLE_CASE( VIL_PIXEL_FORMAT_SBYTE );
+    HANDLE_CASE( VIL_PIXEL_FORMAT_UINT_16 );
+    HANDLE_CASE( VIL_PIXEL_FORMAT_UINT_32 );
+    HANDLE_CASE( VIL_PIXEL_FORMAT_UINT_64 );
```

Comment:
BTW, is there a reason signed int{16,32,64} aren't supported? (I know why floating point aren't; the binning would be effectively impossible!)

Reply:
Good catch, I have included them. FYI, there was initially code to do the calculations on floating points, but we removed it because it seemed unlikely to be used.

Reply:
...Did it look radically different?

I can't imagine how that would work without additional configuration. If you're just using floating point because you are weird, it seems you would need to specify in the configuration what the maximum value is, because it probably *isn't* `limits< double >::max()`. OTOH, if you're working with actual HDR stuff, you probably need some kind of logarithmic binning, because in HDR, 30.0 and 30.5 are "more similar" than 0.1 and 0.4.